named-item hint changes for DDR community

mracsys opened the pull request:
Currently the named-item hint type draws from the item_hints list in the order provided. This worked fine for bingo hints as all items were meant to hinted. Some community members would like to use this to force hints for certain items from a pool. This is implemented by shuffling the item list after the required hints are generated.

Additionally, a new `vague_named_items` key is added to hint distributions to change named item hints to avoid naming the item. Chosen text is "may be on the hero's path" based on community feedback.

Finally, named item hints currently remove entries from the `item_hints` list to avoid double-hinting named items. This had the unintended side effect of removing the item from the spoiler as well. A copy of the named item list is made on world initialization to prevent this. The `hint_dist` setting is also set to `custom` when a plando hint distribution is used to avoid confusion.

Tested with the below plando for different and identical seeds.

```
{
  "settings":            {
    "item_hints": [
        "Boss Key (Forest Temple)",      
        "Boss Key (Fire Temple)", 
        "Boss Key (Water Temple)",
        "Boss Key (Shadow Temple)",
        "Boss Key (Spirit Temple)",
        "Hover Boots",
        "Progressive Hookshot",
        "Dins Fire",
        "Bomb Bag",
        "Boomerang",
        "Bow",
        "Megaton Hammer",
        "Iron Boots",
        "Magic Meter",
        "Mirror Shield",
        "Fire Arrows",
        "Progressive Strength Upgrade",
        "Kokiri Sword" 
    ],
    "hint_dist_user": {
      "name":                  "custom",
      "gui_name":              "Plando Test Hint Distribution",
      "description":           "for debugging",
      "add_locations":         [],
      "remove_locations":      [],
      "add_items":             [],
      "remove_items":          [],
      "dungeons_woth_limit":   2,
      "dungeons_barren_limit": 1,
      "named_items_required":  false,
      "vague_named_items":  true,
      "distribution":          {
          "trial":      {"order":  1, "weight": 0.0, "fixed":   0, "copies": 2},
          "always":     {"order":  2, "weight": 0.0, "fixed":   0, "copies": 0},
          "woth":       {"order":  3, "weight": 0.0, "fixed":   0, "copies": 2},
          "barren":     {"order":  4, "weight": 0.0, "fixed":   0, "copies": 1},
          "entrance":   {"order":  5, "weight": 0.0, "fixed":   0, "copies": 2},
          "sometimes":  {"order":  6, "weight": 0.0, "fixed":   0, "copies": 1},
          "random":     {"order":  7, "weight": 0.0, "fixed":   0, "copies": 2},
          "item":       {"order":  8, "weight": 0.0, "fixed":   0, "copies": 2},
          "song":       {"order":  9, "weight": 0.0, "fixed":   0, "copies": 2},
          "overworld":  {"order": 10, "weight": 0.0, "fixed":   0, "copies": 2},
          "dungeon":    {"order": 11, "weight": 0.0, "fixed":   0, "copies": 2},
          "junk":       {"order": 12, "weight": 0.0, "fixed":   0, "copies": 1},
          "named-item": {"order": 13, "weight": 1.0, "fixed":   0, "copies": 4}
      }
    }
  },
  "randomized_settings": {},
  "starting_items":      {},
  "item_pool":           {},
  "entrances":           {},
  "locations":           {}
}
```

mracsys committed:
Shuffle named-item hints if all item not required to be hinted

mracsys commented:
Reverting to draft to add in additional changes for named-item hints requested by the DDR community.

mracsys committed:
Add vague named item hints and fix spoiler list of candidate named items

mracsys committed:
Change vague item hint wording and fix hint_dist setting in spoilers

mracsys committed:
clarify comment on item_List shuffle

mracsys commented on Hints.py:
```diff
@@ -730,6 +736,7 @@ def buildWorldGossipHints(spoiler, world, checkedLocations=None):
 
         if world.shopsanity != "off" and "Progressive Wallet" not in world.item_hints:
             world.item_hints.append("Progressive Wallet")
+        world.named_item_pool = world.item_hints
```
Good catch, this was intended to be a copy. `item_hints` is the list that ends up in the settings section of the spoiler. To avoid double-hinting an item, I pop it from the list, but this was also removing it from the spoiler. Making a copy lets me mess with the list without affecting the spoiler.

mracsys commented on Hints.py:
```diff
@@ -455,14 +455,20 @@ def get_specific_item_hint(spoiler, world, checked):
     if location.parent_region.dungeon:
         location_text = getHint(location.parent_region.dungeon.name, world.clearer_hints).text
-        return (GossipText('#%s# hoards #%s#.' % (location_text, item_text), ['Green', 'Red']), location)
+        if world.hint_dist_user['vague_named_items']:
```
Which approach would you prefer? `.get()` would be easiest for me.

Zannick replied:
`.get()` is fine, you just have to do it everywhere. I'd prefer defaultdict, but it's not straightforward to do that, so maybe another time.

mracsys replied:
Sounds good. I'll go with `.get()` for now.

mracsys committed:
Address review comments and add item rerolls for named-item hints

An unintended quirk was identified in a race where not all fixed named-item hints were present. The issue was the rolled item was already hinted in an always hint, which was caught in the location filters. To ensure a named-item hint is returned when requested and items still exist in the list, the function now loops until at least one hintable location is available or the item list is exhausted.

Zannick commented on Main.py:
```diff
@@ -104,6 +104,10 @@ def resolve_settings(settings, window=dummy_window()):
 
+    # Set to a custom hint distribution if plando is overriding the distro
+    if len(settings.hint_dist_user) == 0:
+        settings.hint_dist = 'custom'
```
This is apparently causing a crash on my end; I'm guessing because you're doing this in `resolve_settings` which is before `generate`, when we build the World, it looks for a "custom.json" hint dist file, which I don't have. Not sure how to correctly handle it, unless you have it backwards and it should be `!= 0`?

mracsys replied:
Yep, copy/paste error from the World constructor. I don't know why this worked on my side.

mracsys committed:
fix typo for hint dist spoiler output override
